chain-extensions/xvm/src/lib.rs

```diff
@@ -0,0 +1,103 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::pallet_prelude::{Decode, Encode};
+use frame_system::RawOrigin;
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, RetVal, SysConfig, UncheckedFrom,
+};
+use pallet_xvm::XvmContext;
+use sp_runtime::DispatchError;
+use sp_std::vec::Vec;
+
+enum XvmFuncId {
+    XvmCall,
+    // TODO: expand with other calls too
+}
+
+impl TryFrom<u32> for XvmFuncId {
+    type Error = DispatchError;
+
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        match value {
+            1 => Ok(XvmFuncId::XvmCall),
+            _ => Err(DispatchError::Other(
+                "Unsupported func id in Xvm chain extension",
+            )),
+        }
+    }
+}
+
+/// Since we cannot read multiple variable length params, we need to resort to this approach. TODO: can it be improved?
+#[derive(Encode, Decode)]
+struct CustomParams {
+    /// virtual machine identifier
+    vm_id: u8,
```

Comment:
I suggest to use generic here and get precise type from runtime.

Reply:
It could be Decodable.

Reply:
It already is decodable but I don't think this is the right approach given that we should make all types concrete to be used by the smart contract (cannot have `T::VmId`).

So we'll have to move such types to some crate with common smart contract types that are also shared by the chain extensions, to avoid ugly code duplication.